src/main/java/com/google/sps/servlets/GoDataServlet.java

```diff
@@ -14,108 +14,84 @@
 
 package com.google.sps.servlets;
 
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
 import com.google.gson.Gson;
-import com.google.sps.data.HuntItem;
-import com.google.sps.data.LatLng;
-import com.google.sps.data.Riddle;
 import com.google.sps.data.ScavengerHunt;
 import java.io.IOException;
-import java.util.ArrayList;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Returns a fake scavenger hunt and updates the current index of the hunt. */
+/** Returns a scavenger hunt and updates the current index of the hunt. */
 @WebServlet("/go-data")
 public class GoDataServlet extends HttpServlet {
   private static final String INDEX_PARAMETER = "new-index";
 
-  // Keeps track of the current stage in the hunt that the user is on.
-  // Before the user begins the hunt, the index should be -1.
-  private int index = -1;
+  private static final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
   /**
    * Updates the index of the scavenger hunt (aka the destination that the user currently needs to
-   * find.
+   * find).
    */
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Gson gson = new Gson();
     String indexStr = request.getParameter(INDEX_PARAMETER);
+    String huntID = request.getParameter(Constants.HUNTID_PARAMETER);
     try {
-      index = Integer.parseInt(indexStr);
+      int index = Integer.parseInt(indexStr);
+      Entity huntEntity = findScavengerHunt(huntID);
+      if (huntEntity == null) {
+        return;
+      }
+
+      // Update index of scavenger hunt.
+      ScavengerHunt hunt =
+          gson.fromJson((String) huntEntity.getProperty(Constants.HUNT_VAL), ScavengerHunt.class);
+      hunt.updateIndex(index);
+
+      // Convert hunt to JSON string and put into Datastore
+      String huntStr = gson.toJson(hunt);
+      huntEntity.setProperty(Constants.HUNT_VAL, huntStr);
+      datastore.put(huntEntity);
```

Comment:
The main purpose of this code is to update the index, but in order to do so you have to modify the whole hunt_val string, which isn't ideal.

I think in an alternate design hunt_val would be independent of hunt_index so you could update them independently without having to convert back and forth.  I will not make this blocking, but I just wanted to note it here.

Reply:
If I stored hunt_index independently from hunt_val, I think I'd still need to modify the entire hunt_val string in the doGet method, where I send the scavenger hunt to go.js, based on my current design.

Another possibility could be changing the ScavengerHunt class to not include the index, though that'd involve figuring out a way to send both the index and the scavenger hunt in the doGet() response.

Reply:
We chatted about this and brainstormed a way to do this and decided to focus on it next week.  thanks!